### PR TITLE
✨ Listing usb devices on MacOS

### PR DESCRIPTION
**Pull request by muchzill4**
### Description

Listing USB devices on MacOS. This is not hooked up to the cli yet, but working when debugging locally via temp command:
Screenshot 2020-02-13 at 17 10 26

How it works:
1. Establish a list of end usb devices using `system_profiler`.
2. For each device fetch additional data from `ioreg` and `diskutils`.
3. Produce `Candidates`.

Hope you don't mind the size, I'd struggle to chunk this.
I still need to run tests against @acabarbaye giga device hub and some older os versions.

### Test Coverage

- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: mbed_devices/_internal/darwin/diskutil.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        if device.get("DeviceIdentifier") == device_identifier:
+            return device
+    return None
+
+
+def get_mount_point(device_identifier: str) -> Optional[str]:
+    """Returns mount point of a given device."""
+    device_data = get_external_volume_data(device_identifier)
+    if device_data and "MountPoint" in device_data:
+        return cast(str, device_data["MountPoint"])
```
Comment (rwalton-arm) [inline mbed_devices/_internal/darwin/diskutil.py]: I'm not sure why we need this, is it something to do with `mypy`?

File: mbed_devices/_internal/darwin/device_detector.py [hunk 2]
```diff
@@ -1,12 +1,101 @@
...
+
+def _get_serial_port(device_data: system_profiler.USBDeviceData) -> Optional[str]:
+    """Returns serial port for a given device."""
+    device_name = device_data.get("_name")
+    location_id = device_data.get("location_id")
+    if device_name and location_id:
+        ioreg_name = _build_ioreg_device_name(device_name=device_name, location_id=location_id)
+        return ioreg.get_io_dialin_device(ioreg_name)
+
+    return None
```
[the remark this answers is not in the document]
Comment (muchzill4) [reply mbed_devices/_internal/darwin/device_detector.py]: That's a good idea. 

FWIW, the main gripe is that data returned from `system_profiler`, `ioreg` and `diskutil` do not have a schema. I mean, it has some schema, but you can't reliably tell if values are going to be there or not. Hence a bit cowboyish approach, where those schemaless modules return data in raw form (essentially parsed plist values, maybe some filtering) which is then validated here.

File: mbed_devices/_internal/darwin/device_detector.py [hunk 3]
```diff
@@ -1,12 +1,101 @@
...
+            logging.debug(f"Building from: {device_data}.")
+            try:
+                candidate = _build_candidate(device_data)
+            except MissingValues as e:
+                logging.debug(str(e))
+                continue
+            logging.debug(f"Built candidate: {candidate}.")
+            candidates.append(candidate)
+
+        return (c for c in candidates)
```
[the remark this answers is not in the document]
Comment (muchzill4) [reply mbed_devices/_internal/darwin/device_detector.py]: Agreed with all this, I'm not sure why we established `Generator` as a return value here. It was probably YAGNI kind of thing where we'd like to expose lazy evaluation if necessary. Definitely should not bother with that yet - also, nothing lazy happening in MacOS adapter. :)

File: mbed_devices/_internal/darwin/diskutil.py (hunk 1 above)
Comment (muchzill4) [reply mbed_devices/_internal/darwin/diskutil.py]: Yeah. `MountPoint` is always (with a degree of schema-less confidence) a string. I know it, but `mypy` doesn't, as `device_data` is a `Dict` (values can be anything). You can use `cast` to tell `mypy` about the type. I'm not sure if this is the best, but I couldn't find other way to shut it up. :)

File: mbed_devices/_internal/darwin/device_detector.py [hunk 4]
```diff
@@ -1,12 +1,101 @@
...
+    storage_identifiers = [media["bsd_name"] for media in device_media if "bsd_name" in media]
+    mount_points = []
+    for storage_identifier in storage_identifiers:
+        mount_point = diskutil.get_mount_point(storage_identifier)
+        if mount_point:
+            mount_points.append(pathlib.Path(mount_point))
+    return mount_points
+
+
+class MissingValues(Exception):
```
[the remark this answers is not in the document]
Comment (muchzill4) [reply mbed_devices/_internal/darwin/device_detector.py]: I've updated the base class, but unsure if specific exceptions are worth it. Tried to give informative strings to an exception class I've added. Let me know your thoughts.

File: mbed_devices/_internal/darwin/device_detector.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (muchzill4) [reply mbed_devices/_internal/darwin/device_detector.py]: Updated, code now raises exceptions and avoids `None` where we can.

**Comment by muchzill4** [comment]
FWIW, I started working on the code higher up, that'd use this adapter and realized that `Candidate` can essentially be half `None` and would still be interrogated to see if it's a valid `MbedDevice`.
Determining if something is a valid `MbedDevice` happens by applying various flows:
- checking `vendor_id`, `product_id` 
- reading the files in the `mount_points`,
- etc.
I think it should not be adapters responsibility to perform validations like that or any pre-validation for that matter (check if candidate has _any_ values). It's potentially expensive to be completely true and will be performed higher up the stack anyway.

File: setup.py [hunk 5]
```diff
@@ -34,7 +34,7 @@
     description="This package allows to list Mbed OS devices which are connected to the host computer.",
     keywords="Arm Mbed OS MbedOS devices list",
     include_package_data=True,
-    install_requires=["python-dotenv", "Click==7.0", "mbed-targets"],
+    install_requires=["python-dotenv", "Click==7.0", "mbed-targets", "mbed-tools-lib"],
```
Comment (acabarbaye) [inline setup.py]:
```suggestion
    install_requires=["python-dotenv", "Click==7.0", "mbed-targets"],
```
Could you remove this for now as it is not a dependency yet?

File: mbed_devices/_internal/darwin/system_profiler.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    for device in data:
+        if "_items" in device:
+            child_devices = _filter_end_devices(device["_items"])
+            end_devices.extend(child_devices)
+        else:
+            end_devices.append(device)
+    return [device for device in end_devices if _is_end_device(device)]
+
+
+def _is_end_device(data: USBDeviceData) -> bool:
```
Comment (acabarbaye) [inline mbed_devices/_internal/darwin/system_profiler.py]: Could you add a comment of what you mean by end device

Comment (muchzill4) [reply mbed_devices/_internal/darwin/system_profiler.py]: Fixed by writing more explicit code.

**Pull request merged**